File: src/algorithms/arrays/counting-sort/sources/counting-sort.ts
// Counting Sort — O(n+k) sort by counting frequencies and reconstructing sorted order
function countingSort(inputArray: number[]): number[] {
  if (inputArray.length === 0) {
    // @step:initialize
    return []; // @step:initialize
  }

  const maxValue = Math.max(...inputArray); // @step:initialize
  const countArray: number[] = new Array(maxValue + 1).fill(0); // @step:initialize

  // Count the frequency of each element
  for (let scanIndex = 0; scanIndex < inputArray.length; scanIndex++) {
    countArray[inputArray[scanIndex]!]!++; // @step:visit
  }

  // Reconstruct the sorted array from count frequencies
  const sortedArray: number[] = []; // @step:compare
  for (let currentValue = 0; currentValue <= maxValue; currentValue++) {
    for (let repeatIndex = 0; repeatIndex < countArray[currentValue]!; repeatIndex++) {
      sortedArray.push(currentValue); // @step:compare
    }
  }

  return sortedArray; // @step:complete
}

File: src/algorithms/arrays/counting-sort/sources/CountingSort.java
// Counting Sort — O(n+k) sort by counting frequencies and reconstructing sorted order
public class CountingSort {
    public static int[] countingSort(int[] inputArray) {
        if (inputArray.length == 0) { // @step:initialize
            return new int[0]; // @step:initialize
        }

        int maxValue = inputArray[0]; // @step:initialize
        for (int scanIndex = 0; scanIndex < inputArray.length; scanIndex++) {
            if (inputArray[scanIndex] > maxValue) maxValue = inputArray[scanIndex];
        }
        int[] countArray = new int[maxValue + 1]; // @step:initialize

        // Count the frequency of each element
        for (int scanIndex = 0; scanIndex < inputArray.length; scanIndex++) {
            countArray[inputArray[scanIndex]]++; // @step:visit
        }

        // Reconstruct the sorted array from count frequencies
        int[] sortedArray = new int[inputArray.length]; // @step:compare
        int outputIndex = 0;
        for (int currentValue = 0; currentValue <= maxValue; currentValue++) {
            for (int repeatIndex = 0; repeatIndex < countArray[currentValue]; repeatIndex++) {
                sortedArray[outputIndex++] = currentValue; // @step:compare
            }
        }

        return sortedArray; // @step:complete
    }
}
